eng/common/scripts/job-matrix/job-matrix-functions.ps1

```diff
@@ -365,18 +367,46 @@ function ProcessReplace
     return $replaceMatrix
 }
 
-function ProcessImport([MatrixParameter[]]$matrix, [String]$selection, [Array]$nonSparseParameters, [Hashtable]$displayNamesLookup)
-{
+function ProcessEnvironmentVariableReferences([array]$matrix, $displayNamesLookup) {
+    $updatedMatrix = @()
+
+    foreach ($element in $matrix) {
+        $updated = [MatrixParameter[]]@()
+
+        foreach ($perm in $element._permutation) {
+            # Iterate nested permutations or run once for singular values (int, string, bool)
+            foreach ($flattened in $perm.Flatten()) {
+                if ($flattened.Value?.GetType() -eq "".GetType() -and $flattened.Value.StartsWith("env:")) {
+                    $envKey = $flattened.Value.Replace("env:", "")
+                    $value = [System.Environment]::GetEnvironmentVariable($envKey) ?? ""
+                    if (!$value) {
+                        Write-Warning "Environment variable `"$envKey`" was not found or is empty."
+                    }
+                    $flattened.Set($value, $flattened.Name)
+                }
+            }
+
+            $updated += $perm
+        }
+
+        $updatedMatrix += CreateMatrixCombinationScalar $updated $displayNamesLookup
+    }
+
+    return $updatedMatrix
+}
+
+function ProcessImport([MatrixParameter[]]$matrix, [String]$selection, [Array]$nonSparseParameters, [Hashtable]$displayNamesLookup) {
     $importPath = ""
     $matrix = $matrix | ForEach-Object {
         if ($_.Name -ne $IMPORT_KEYWORD) {
             return $_
-        } else {
+        }
+        else {
             $importPath = $_.Value
         }
     }
     if ((!$matrix -and !$importPath) -or !$importPath) {
-        return $matrix, @(), @{}
+        return $matrix, @(), $displayNamesLookup
```

Comment:
Also fixed a separate bug here where display names lookup was getting lost if we didn't have imports